R/card.R

```diff
@@ -0,0 +1,298 @@
+#'   known card items (like [card_header()], [card_plot()], etc.) and if not,
+#'   they are automatically wrapped in [card_body()]. If `autowrap` is `FALSE`,
+#'   then all unnamed arguments are nested directly within the card element with
+#'   no further processing.
+#'
+#' @export
+card <- function(..., class = NULL, width = NULL, height = NULL,
+  autowrap = TRUE) {
+
+  width <- htmltools::validateCssUnit(width)
+  height <- htmltools::validateCssUnit(height)
+
+  args <- rlang::list2(...)
+  argnames <- names(args)
+  if (is.null(argnames) && length(args) > 0) {
+    argnames <- rep_len("", length(args))
+  }
+
+  attribs <- args[nzchar(argnames)]
+  children <- args[!nzchar(argnames)]
+
+  if (isTRUE(autowrap)) {
+    # Any children that are `is.card_item` should be included verbatim. Any
+    # children that are not, should be wrapped in card_body(). Consecutive children
+    # that are not card_item, should be wrapped in a single card_body().
+    needs_wrap <- !vapply(children, is.card_item, logical(1))
+    needs_wrap_rle <- rle(needs_wrap)
+    start_indices <- c(1, head(cumsum(needs_wrap_rle$length) + 1, -1))
+    children <- mapply(start_indices, needs_wrap_rle$length, needs_wrap_rle$values,
+      FUN = function(start, length, wrap) {
+        these_children <- children[start:(start + length - 1)]
+        if (wrap) {
+          list(card_body(stretch = FALSE, these_children))
+        } else {
+          these_children
+        }
+      }, SIMPLIFY = FALSE)
+    children <- unlist(children, recursive = FALSE)
+  }
+
+  tag <- htmltools::tags$div(
+    class = "card",
+    class = class,
+    style = htmltools::css(width = width, height = height),
+    !!!attribs,
+    !!!children
+  )
+  as_fragment(
+    tag_require(tag, version = 4, caller = "card()")
+  )
+}
+
+#' @rdname card_body
+#' @export
+as.card_item <- function(x) {
+  class(x) <- c("card_item", class(x))
+  x
+}
+
+#' @rdname card_body
+#' @export
+is.card_item <- function(x) {
+  inherits(x, "card_item")
+}
+
+#' Card components
+#'
+#' @description This topic describes various components that are intended to go
+#'   directly inside of a [card()]. These components can be used in combination;
+#'   for example, a single card could contain a `card_header()`, multiple
+#'   `card_body()`s, a `card_list()`, and finally a `card_footer()`.
+#'
+#' @param ... Named arguments become attributes on the `<div class="card">`
+#'   element. Unnamed arguments become card items, and can be any valid child of
+#'   an [htmltools tag][htmltools::tags].
+#' @param class Additional CSS classes to include on the card div.
+#' @param padding Whether vertical and/or horizontal padding should be included.
+#'   Use `NULL` for neither, `"x"` for horizontal only, `"y"` for vertical only,
+#'   or `c("x", "y")` for both (the default).
+#' @param stretch Set to `TRUE` if this `card_body` is eager to use any extra
+#'   vertical space is available in the card.
+#'
+#' @export
```

Comment:
We should also include `@example` and/or link to the article

```suggestion
#' @export
#' @family cards
```